src/vehicle_autodrive.cpp

```diff
@@ -1,25 +1,1202 @@
+ * based on proximity to the "exit zone" leading to the next OMT and whether the vehicle is
+ * pointing towards the middle of that zone). Although there can be many paths that we could
+ * take to get through to the next OMT, we just take the first one we find since it tends to
+ * be the most direct and is usually near-optimal. When the way is blocked we'll either give
+ * up after exhausting the search space or after we hit a limit on the search length.
+ *
+ * In order to build the search graph for A*, we consider every possible map tile and orientation
+ * in the search zone and we figure out whether the vehicle's pivot point could be placed there
+ * without the vehicle overlapping some obstacle (a collision-free placement). The pivot point
+ * is the part of the vehicle that moves the least while steering and is usually between the
+ * wheels of the vehicle, closer to the non-steerable ones. The pathfinding problem reduces to
+ * finding a path for the pivot point to follow through the set of possible positions and
+ * orientations (nodes). In order to figure out the possible connections between nodes (the
+ * edges of the graph) we consider a limited set of actions that the driver can perform at
+ * each location: nothing, steer left or steer right (possibly more than once). Given a current
+ * node, current speed and amount of steering we can predict where the vehicle's pivot point
+ * will end up in 1 turn and connect the 2 nodes.
+ *
+ * In order to keep the search space small we only consider paths traveled at constant speed
+ * (otherwise we'd have to add speed as an extra attribute in every node). When we need to
+ * compute a new path we first try to use the maximum safe autodrive speed (which is typically
+ * much less than the vehicle's max speed). If we can't find a path we cut the speed in half and
+ * try again, repeating until we succeed or reach the minimum speed. When the target speed is
+ * different from the current speed of the vehicle we try to estimate how long it will take to
+ * reach that speed and take that into account when simulating the first few edges (they'll be
+ * shorter or longer than the rest).
+ *
+ * Once we have generated a sequence of nodes to follow, we do not invoke A* again until we reach
+ * the next OMT. We simply cache the path and on every subsequent call to do_autodrive() we
+ * check where we are on that path, what node we're supposed to reach next and what action we
+ * should use to get there. However, sometimes the simplified simulation used to compute the
+ * search graph fails to accurately predict where the vehicle will end up. This could be because
+ * there was too much error in an approximation (such as time to accelerate) or the driving
+ * conditions changed (ran over a mound of dirt that interfered with steering). When such a
+ * discrepancy is detected, autodrive will fall back to "safe mode", cutting the speed to the
+ * minimum (1 tile per second) and recomputing the path.
```

Comment:
Note the possibility of hooking code into the avatar "safe mode" to cancel autodriving if that goes off. (A different name for the autodrive "safe mode" may be a good idea.)

Reply:
I'm not actually calling it safe mode in the code. I can change the comment if it's confusing.
All I'm trying to say is that when simulation errors are detected we reduce the speed to 1 tile/s. If we get simulation errors at that speed, then we abort.

Reply:
Whoops; already does happen (re avatar safe mode) - never mind.

Reply:
Activities are actually not affected by "SAFE mode". That game feature only blocks player-initiated motion (e.g. by pressing the arrow keys) when it triggers.
Activities have their own independent safety feature. When you see a zombie while autodriving you'll see:
![image](https://user-images.githubusercontent.com/8000047/126576091-0b3ca07a-2a0c-406c-bfac-410bebd2a11b.png)
If you ignore it you'll get another warning when an enemy gets "dangerously close".
The same thing happens while reloading a gun, chopping a tree, crafting, etc.